man/md5test.1.man
.\" Copyright (c) 2001 Francesco Paolo Lovergine <francesco@yardradius.org>
.\" See copyright section for conditions of rednistribution
.\" $Id: md5test.1.man,v 1.5 2004/07/01 13:57:55 flovergine Exp $
.\"
.TH MD5TEST 1 "Aug 28, 2001" "1.0.19" "Yard Radius Manual"
.SH NAME
md5test \- MD5 test program for RADIUS
.SH SYNOPSIS
.BI md5test " shared"
.SH DESCRIPTION
.B md5test
is a poorly written program for testing purpose.
It takes the shared secret as an argument,
reads a hex dump on standard input, and outputs
the length, shared secret, standard input, followed
by the MD5 checksum of that input followed by the
shared secret.
.SH "SEE ALSO"
.BR radiusd (8),
.BR md5sum (1)
.SH BUGS
No decent argument and input checking is done. Use it with caution, due 
to easy segmentation faulting.
.SH AUTHOR
Lucent Technologies Remote Access.
.SH COPYRIGHT
Copyright (C) 1992-1999 Lucent Inc. All rights reserved.
.P
Copyright (C) 1999-2001 Francesco Paolo Lovergine. All rights reserved.
.P
See the LICENSE file enclosed within this software for conditions of 
use and distribution. This is a pure 
.B "ISO BSD Open Source License" .



ChangeLog
[ $Id: ChangeLog,v 1.29 2004/07/01 13:57:52 flovergine Exp $ ]

(1.0.21) 2004-07-01 Francesco P. Lovergine <francesco@yardradius.org>

	The 'middle-land' release.

	  [Minor]
	 
	* Editorial changes in copyright notices of all source files.
	  Also some typos are now corrected.
	* New email and domain used extensively in every file.
	* Updated README to reflect new site URL and other info.
	* Refinement in size definitions of some data structures.

	  [Enhancement]
	  
	* Changed --version in -dumpversion for gcc versioning 
          in configure.in. This broke things in gcc 3. New arg is
          back compatible also.
	* Added entries limit to radlast. Thanks Russell Heilling 
	  for patch (see feature request 464791). Changed man page
	  as consequence.
	* Updated config.guess and config.sub, old releases could
	  break thing on current systems.

	  [Security]

	* A buffer overflow in calc_acctreq() has been patched.
	* process_menu() has been modified to check string limits.
	  Thanks Max Vozeler for pointing this so old issue and
	  showed a possible exploit.

(1.0.20) 2002-03-02 Francesco P. Lovergine <francesco@yardradius.org>
	* Documentation has been revised and updated. Many typos are
	  now corrected as well as some documentation incoherences.
	  New RFCs about the RADIUS protocol suite are enclosed and
	  obsolete ones removed.
	* Some files changed names for coherence.
	* getopt() returns int, not char.
	* SIGABRT is the standard definition in POSIX.1 which obsoletes
	  SIGIOT. In some environment the old signal is no more defined.
	* Now changeprefix script works good on Free BSD. The non
	  portable '-' argument has been skipped for sed. This caused
	  problem with configure script under that platform.
	* Dictionary updated with other VSAs (Ericsson and Xedia).
	* config.sub, config.guess and other automake scripts are now 
	  updated to the latest releases.
	* Major code revisions to avoid vulnerabilities due to strcpy()
          memcpy() and other dangerous functions. 
           - Digest Calculation Buffer Overflow Vulnerability.
           - Invalid attribute length calculation on malformed VSA.
          Non full-disclosures are available at CERT.org
	* Added my GPG key to code.

	  Some of those changes are stolen from a VA Linux patch for
	  Lucent Radius 2.1.

(1.0.19) 2001-08-27 Francesco P. Lovergine <francesco@yardradius.org>
	* Added a first version of man files under `man' directory.
	  Some old draft documentation files have been removed,
	  as consequence.
	* Default prefix is changed in `/usr/local/yardradius'
	* A couple of commas in configure.in which caused a ',6' 
	  file when --enable-debian was used has been deleted. 
	  A minor, but truly amazing bug :)
	* Many typos and grammar errors in doc corrected (or hoping so).
	* Configure.in did not install logs and doc directory when
	  prefix is not set on command line. Now also this is ok.

(1.0.18) 2001-08-21 Francesco P. Lovergine <francesco@yardradius.org>
	* Patched for a subdle bug due to __debug definition in extens.c.
	* Skipped auth_item check when Auth-Type=PAM.
	* Skipping void reply items in parsing GDBM users files.

	  All patches are due to William K. Volkman.

(1.0.17) 2001-08-18 Francesco P. Lovergine <francesco@yardradius.org> 
	* Configuration file modified to add a --enable-debian option.
	  It greatly simplifies support of Debian packaging. Debian
	  package is released as a separate package in Debian 'unstable'
	  distribution.
	* Now builddbm and radiusd are installend under sbin which
	  seems a more correct destination for those files.

(1.0.16/pre16) 2001-01-14 Francesco P. Lovergine <francesco@yardradius.org> 
	* Dictionary entries for BayNetworks/Nortel Annex is added.
	* Now, YARD RADIUS is compatible with GDBM 1.7.3, which seems the 
	  most commonly installed version. Just changed the gdbm_open()
	  function arguments for this.
	* Since now, the official name of the package is
	  Yard Radius. This to avoid confusion with other packages
	  with the same name, and to be coherent with a next-to-come 
	  Debian package (yeah!).
	  I changed also version numbering. The current version is 
	  1.0.16, i.e. 1.0 patch level 16. It seems more `standard',
	  and to be honest I really think this program is too much
	  robust to leave it in pre-release status... Am I an optimist?
	* Changed including of time.h and sys/time.h as suggested in autoconf
	  manuals. This allows installation under AIX and other systems
	  which could need both files. Incidentally, under AIX it compiles
	  also with the off-the-shelf C compiler.
	* PAM configuring in configure.in is changed to avoid a 
	  subdle configuration bug: some linux distributions do not
	  install PAM development libraries, so that PAM support cannot
	  be enabled on them, while building. This includes Debian 2.2
	  and RedHat.  That's not a problem on Solaris or FreeBSD 
	  which install them as default. This require a modification
	  of configure.in which is now smarter. 
	* An silly ortographic error in configure.in with --enable-pam
	  has been corrected.
	* Builddbm now works under the conf directory, not the current one.
	* License and Readme files have been changed. There is no more
	  an ANNOUNCE file and COPYING (which is the standard name for
	  GPL) is now named LICENSE (which is the standard name for BSD).
	* A CONFIGURATION text file has been added to summarize main
	  configuration issues. Are you happy?
	* Many other (little) adjustments.
	
(pre15) 2000-09-26 Francesco P. Lovergine <francesco@yardradius.org>
	* Installer now create also logs directory and explains
	  the structure.
	* Deleted Log entries from sources to avoid redundant comments
	  with CVS.

(pre14) 2000-09-21 Francesco P. Lovergine <francesco@yardradius.org>
	* Corrected autoconf/automake 
	  configuration to solve some problems of installation
	  under a few architectures. Configure now use Gnu Shell Tool

src/md5test.c
/*
 *	md5test - calculate md5 checksum for testing purposes
 */

/***********************************************************************

RADIUS
Remote Authentication Dial In User Service

Lucent Technologies Remote Access
4464 Willow Road
Pleasanton, CA   94588

Copyright 1992-1999 Lucent Technologies Inc.  All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are
met:

   * Redistributions of source code must retain the above copyright
     notice, this list of conditions and the following disclaimer.

   * Redistributions in binary form must reproduce the above
     copyright notice, this list of conditions and the following
     disclaimer in the documentation and/or other materials provided
     with the distribution.

   * All advertising materials mentioning features or use of this
     software must display the following acknowledgement:

	  This product includes software developed by Lucent
	  Technologies and its contributors.

   * Neither the name of the copyright holder nor the names of its
     contributors may be used to endorse or promote products derived
     from this software without specific prior written permission.

This software is provided by the copyright holders and contributors
``as is'' and any express or implied warranties, including, but not
limited to, the implied warranties of merchantability and fitness for a
particular purpose are disclaimed. In no event shall the copyright
holder or contributors be liable for any direct, indirect, incidental,
special, exemplary, or consequential damages (including, but not
limited to, procurement of substitute goods or services; loss of use,
data, or profits; or business interruption) however caused and on any
theory of liability, whether in contract, strict liability, or tort
(including negligence or otherwise) arising in any way out of the use
of this software, even if advised of the possibility of such damage.

************************************************************************/


static char rcsid[] =
"$Id: md5test.c,v 1.2 2000/09/07 13:36:28 kiavik Exp $ Copyright 1999 Lucent Technologies Inc";

/*	Usage:	md5test takes the shared secret as an argument,
		reads a hex dump on standard input, and outputs
		the length, shared secret, standard input, followed
		by the MD5 checksum of that input followed by the
		shared secret.

		This skeleton program is useful combined with 
		radiusd -x -x to make sure that the right checksums are
		being calculated

 */

#include <stdio.h>
#include <sys/types.h>

main(argc,argv)
int argc;
char **argv;
{
	u_char buf[256],pw_digest[16];
	int h;
	int i,n = 0,len;

	*argv++;
	while (scanf("%2x",&h) != EOF) {
		buf[n++] = h & 0xff;
	}
	printf("%d %s\n",n,*argv);
	len=strlen(*argv);
	memcpy(&buf[n],*argv,len);
	md5_calc((u_char*)pw_digest, (u_char*)buf, n+len);
	for (i=0;i<n;i++) {
		printf("%02x ",buf[i]);
		if ((i&0xf) == 0xf) {
			printf("\n");
		}
	}
	printf ("\n");
	for (i=0;i<16;i++) {
		printf("%02x ",pw_digest[i]);
	}
	printf ("\n");
}

